admin/class-acc_user_importer-admin.php

```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
+
+					$this->log_dual(" > existing user #" . $user_id . " " . $username2 .
+					                " already has email " . $user_email);
+					$this->log_dual(" > user update skipped");
 					$user_data = array_merge( array('ID' => $user_id), $user_data); //point to correct ID (future proofing)
-					$update_this_user = false;
 					$update_errors[] = $user;
-				}
-				
-				//no duplicate email exists
-				else {
-					$api_response['log'] .= "&nbsp;&gt; email not found on any other users<br/>";
-					$api_response['log'] .= "&nbsp;&gt; creating new user account<br/>";
-					$this->log_local_output(" > email not found on any other users");
-					$this->log_local_output(" > creating new user account");
+				} else {
+					//email is unique, proceed to create a new record
+					$update_this_user = true;
+					$create_this_user = true;
+					$this->log_dual(" > email not found on any other users");
+					$this->log_dual(" > will create new user account");
 					$new_users[] = $user_contact_id;
+					$new_users_email[] = $user_email ;
 					$user_data["role"] = $default_role;
 				}
 			}
 			
 			//didn't find a user by contact ID, and don't have an email to search for either
 			else {
-				$api_response['log'] .= "&nbsp;&gt; <b>error</b>: no email given, cannot create new user account.<br/>";
-				$this->log_local_output(" > error: no email given, cannot create new user account.");
-				$update_this_user = false;
+				$this->log_dual(" > error: no email given, cannot create new user account.");
 				$update_errors[] = $user['FirstName'] . " " . $user['LastName'] . " (" . $user_contact_id . ")";
 			}
 			
+			//If user already exists, update only if something changed.
+			//Do a bunch of compare to decide if an update is needed.
+			//This logic needs to be undated each time we add a new user field.
+			if ($update_this_user && $user_already_exists) {
 
-			//only run if indicated
-			if ($update_this_user) {
+				$update_this_user = false;
 
-				$wp_user = wp_insert_user( $user_data ) ;
-				
-				//attempt to add user
-				if ( is_wp_error($wp_user) ) {
-					
-					$api_response['log'] .= "&nbsp;&gt; failed to update user<br/>";
-					$api_response['log'] .= "&nbsp;&gt; WP:" . $wp_user->get_error_message() . "<br/>";
-					$this->log_local_output(" > failed to update user");
-					$this->log_local_output(" > WP:" . $wp_user->get_error_message());
+				// Get the user object
+				$user_meta = get_userdata($user_id);
+
+				// Get all the user roles as an array.
+				$user_roles = $user_meta->roles;
+				// Check if user has expired or ex-member role. If so, update role to default
+				if ( in_array( $role_expiry1, $user_roles, true ) ) {
+					$this->log_dual(" > User is " . $role_expiry1 . ", refreshing role to " . $default_role);
+					$update_this_user = true;
+					$user_data["role"] = $default_role;
+					$role_refreshed[] = $user_contact_id;
+					$role_refreshed_email[] = $user_email;
```

Comment:
I have 2 arrays (roles_refreshed, role_refreshed_email) for storing the names and emails of those we refreshed roles.  I have a feeling there is a better way to do it. Feel free to improve.